build.gradle

```diff
@@ -22,6 +22,10 @@ plugins {
     id 'java-base'
     id 'build-types'
     id 'gradle-compile'
+    // TODO Apply this plugin in the BuildScanConfigurationPlugin once binary plugins can apply plugins via the new plugin DSL
+    // We have to apply it here at the moment, so that when the build scan plugin is auto-applied via --scan can detect that
+    // the plugin has been already applied. For that the plugin has to be applied with the new plugin DSL syntax.
+    id "com.gradle.build-scan"
```

Comment:
Why do we now always apply the build scan plugin? While I like to apply it always, we currently have been using an init script to apply the right version in CI. Would we now retire that init-script (`buildScanInit.gradle`)?

Reply:
We would not retire it as the init script still serves the purpose of configuring the behavior (e.g. define the destination of the scan-server). Keep in mind that when `--scan` is used, the behavior is different (send a build scan to scans.gradle.com). 

@ldaley How would we inject a different version (e.g. the latest one to publish to scans.grdev.net) with the new mechanism?

Reply:
We can not as the plugin needs to be in the classpath when the plugin is applied in the root project. Keep in mind that the root project applies the plugin without a version number. This means it has to be in the classpath. It will not provision it from anywhere.

Reply:
I think I understand now.
I tried moving the plugin dependency to `compileOnly` and then apply the `build-scan` plugin with version in the plugins block in `build.gradle`. That yields a `NoClassDefFoundException` when trying to call `getBuildScan()` in `BuildScanConfigurationPlugin.kt`.

That basically means that classes from plugins applied by version in the root build are not available to plugins defined in `buildSrc`. Therefore, the current solution is to add the plugin to the classpath of the main build by adding it to the `buildSrc` classpath.

Sadly, that breaks the Gradle build when I use my init script which adds the build scan plugin to all the projects I am running, see https://e.grdev.net/s/22pagt5asbn26:
```
initscript {
    repositories {
        maven {
            url "https://plugins.gradle.org/m2"
        }
    }
    dependencies {
        classpath "com.gradle:build-scan-plugin:1.12.1"
    }
}

rootProject { prj ->
   apply plugin: initscript.classLoader.loadClass("com.gradle.scan.plugin.BuildScanPlugin")
   buildScan {
       server = "https://e.grdev.net"
       publishAlways()
   }
}
```

Any ideas how I can use an init script like that + have the custom values we add for the Gradle build from a Kotlin DSL script? Maybe @bamboo or @eskatos have some ideas.

Maybe having a Gradle property which enables build scans for all builds would be a workaround. E.g. one would set `org.gradle.scans=true` in the Gradle user home `gradle.properties` and that would imply `--scan` for all builds.

Reply:
@wolfs @eskatos: @ldaley has come up with a solution: 
```
// Apply the build scan plugin if it's not already present
// Works if scan plugin is added via buildSrc, root project plugins {}, or root project 
initscript {
    repositories {
        maven {
            url "https://plugins.gradle.org/m2"
        }
    }
    dependencies {
        classpath "com.gradle:build-scan-plugin:1.12.1"
    }
}

// Will not work reliably if scan plugin is added via a script plugin's dedicated loader 
if (gradle.parent == null) {
  rootProject { prj -> 
    prj.buildscript.configurations.classpath.incoming.afterResolve { // wait until this to be after default plugins
      def first = true
      prj.plugins.whenPluginAdded { plugin -> // wait until this to be after plugin {} application
        if (first) {
          first = false
          try {
            prj.buildscript.classLoader.loadClass("com.gradle.scan.plugin.BuildScanPlugin")
          } catch (ClassNotFoundException e) {
            // Scan plugin not visible to the root project, assume its not coming and apply ours
            prj.apply(plugin: com.gradle.scan.plugin.BuildScanPlugin)
          }  
        }
       
        if (plugin.class.name == "com.gradle.scan.plugin.BuildScanPlugin") {
          prj.buildScan {
            server = "https://e.grdev.net"
            publishAlways()
          }
        }
      }
    }
  }
}
```

Keep in mind that this is very much a corner case. We want to be able to not apply the plugin, apply it via `--scan` and apply it via a generic init script (the above). This is an unusual combination. Organizations usually don't need to deal with all of the above. So I'm happy we have a solution. Many thanks Luke.

For me this was the last block on getting this merged. The GE build is offering a configuration option to switch between the latest snapshot version of the plugin and the latest released one. That would be nice to have but is not blocking a merge. If people want to experiment with other versions of the build scan plugin, they can change the version in the build script of buildSrc.

Reply:
Since those changes I cannot apply the build scan recipes plugin anymore. Since I was probably the only one using it, it didn't get noticed. I spent 15 minutes trying to fix it, but always had different errors (`buildScan` not available, duplicate classes, class cast exception, ...) so I gave up.